src/packaging/licenses/__init__.py

```diff
@@ -0,0 +1,126 @@
+# INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+# PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+# CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
+# OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+#
+# With additional allowance of arbitrary `LicenseRef-` identifiers, not just
+# `LicenseRef-Public-Domain` and `LicenseRef-Proprietary`.
+#
+#######################################################################################
+from __future__ import annotations
+
+import re
+from typing import cast
+
+from packaging.licenses.spdx import EXCEPTIONS, LICENSES
+
+license_ref_allowed = re.compile("^[A-Za-z0-9.-]*$")
+
+
+def normalize_license_expression(raw_license_expression: str) -> str | None:
+    if not raw_license_expression:
+        return None
+
+    license_refs = {
+        ref.lower(): "LicenseRef-" + ref[11:]
+        for ref in raw_license_expression.split()
+        if ref.lower().startswith("licenseref-")
+    }
+
+    # First normalize to lower case so we can look up licenses/exceptions
+    # and so boolean operators are Python-compatible
+    license_expression = raw_license_expression.lower()
+
+    # Then pad parentheses so tokenization can be achieved by merely splitting on
+    # white space
+    license_expression = license_expression.replace("(", " ( ").replace(")", " ) ")
+
+    # Now we begin parsing
+    tokens = license_expression.split()
+
+    # Rather than implementing boolean logic we create an expression that Python can
+    # parse. Everything that is not involved with the grammar itself is treated as
+    # `False` and the expression should evaluate as such.
+    python_tokens = []
+    for token in tokens:
+        if token not in {"or", "and", "with", "(", ")"}:
+            python_tokens.append("False")
+        elif token == "with":
+            python_tokens.append("or")
+        elif token == "(" and python_tokens and python_tokens[-1] not in {"or", "and"}:
+            message = f"invalid license expression: {raw_license_expression}"
+            raise ValueError(message)
+        else:
+            python_tokens.append(token)
+
+    python_expression = " ".join(python_tokens)
+    try:
+        result = eval(python_expression)
+    except Exception:
+        result = True
+
+    if result is not False:
+        message = f"invalid license expression: {raw_license_expression}"
+        raise ValueError(message) from None
+
+    # Take a final pass to check for unknown licenses/exceptions
+    normalized_tokens = []
+    for token in tokens:
+        if token in {"or", "and", "with", "(", ")"}:
+            normalized_tokens.append(token.upper())
+            continue
+
+        if normalized_tokens and normalized_tokens[-1] == "WITH":
```

Comment:
I would recommend against this one, this repository doesn't appear to have many stylistic rules enabled but in Hatch (where this came from) I enable the one where control flow statements require no `elif`. In this example, the `continue` is the statement so the following branch should start anew.

Reply:
I know it isn't necessary, but I have been bit by people not noticing it wasn't an `elif` when changing something like this to not short-circuit the overall `if` block.

Anyway, I'm guessing you care because you want to copy the code back into Hatch?

Reply:
I'm going to use whatever lands here and get rid of my custom code now that it's upstreamed. I just wanted maintainers to be aware that when you start using Ruff more it will catch this line, and I think for good reason. Feel free to do whatever you think is best!

Reply:
I'm a fan of the simpler structure; the `el` isn't necessary, and adds extra complexity (especially when nested). Changing a control flow statement into a non-control flow statement should change control flow, trying to proactively protect against it isn't helpful, IMO.

And this is called a "guard pattern" - it's even part of the syntax in Ruby. (It would be `continue if normalized_tokens and normalized_tokens[-1] == "WITH"` if Python had Ruby's guard, btw).